manager/integration/tests/test_zone.py

```diff
@@ -301,3 +330,611 @@ def test_replica_zone_anti_affinity(client, core_api, volume_name, k8s_node_zone
+    set_k8s_node_zone_label(core_api, n3.name, ZONE3)
+    wait_longhorn_node_zone_updated(client)
+
+    client.update(n2, allowScheduling=False)
+    client.update(n3, allowScheduling=False)
+
+    n_replicas = 6
+    volume = create_and_check_volume(client, volume_name,
+                                     num_of_replicas=n_replicas)
+    volume.attach(hostId=n1.name)
+
+    for _ in range(RETRY_COUNTS):
+        z1_r_count = get_zone_replica_count(
+            client, volume_name, ZONE1, chk_running=True)
+        z2_r_count = get_zone_replica_count(
+            client, volume_name, ZONE2, chk_running=True)
+        z3_r_count = get_zone_replica_count(
+            client, volume_name, ZONE3, chk_running=True)
+
+        if z1_r_count == 6 and z2_r_count == z3_r_count == 0:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 6
+    assert z2_r_count == 0
+    assert z3_r_count == 0
+
+    client.update(n2, allowScheduling=True)
+
+    for _ in range(RETRY_COUNTS):
+        z1_r_count = get_zone_replica_count(
+            client, volume_name, ZONE1, chk_running=True)
+        z2_r_count = get_zone_replica_count(
+            client, volume_name, ZONE2, chk_running=True)
+        z3_r_count = get_zone_replica_count(
+            client, volume_name, ZONE3, chk_running=True)
+
+        if z1_r_count == z2_r_count == 3 and z3_r_count == 0:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 3
+    assert z2_r_count == 3
+    assert z3_r_count == 0
+
+    client.update(n3, allowScheduling=True)
+
+    for _ in range(RETRY_COUNTS):
+        z1_r_count = get_zone_replica_count(
+            client, volume_name, ZONE1, chk_running=True)
+        z2_r_count = get_zone_replica_count(
+            client, volume_name, ZONE2, chk_running=True)
+        z3_r_count = get_zone_replica_count(
+            client, volume_name, ZONE3, chk_running=True)
+
+        if z1_r_count == z2_r_count == z3_r_count == 2:
+            break
+
+        time.sleep(RETRY_INTERVAL)
+    assert z1_r_count == 2
+    assert z2_r_count == 2
+    assert z3_r_count == 2
+
+    reset_longhorn_node_zone(client, core_api)
+
+
+def test_replica_auto_balance_zone_best_effort_with_data_locality(
+        client, core_api, volume_name, pod):  # NOQA
+    """
+    Background:
+    Given set `replica-soft-anti-affinity` to `true`.
+    And set `replica-zone-soft-anti-affinity` to `true`.
+    And set `default-data-locality` to `best-effort`.
+    And set `replicaAutoBalance` to `best-effort`.
+    And set node-1 to zone-1.
+        set node-2 to zone-1.
+        set node-3 to zone-2.
+    And create volume with 2 replicas.
+    And create pv for volume.
+    And create pvc for volume.
+
+    Scenario: replica auto-balance zones with best-effort should not
+              remove pod local replicas when data locality is enabled
+              (best-effort).
+
+    Given create and wait pod on <node>.
```

Comment:
`<node>`is confusing. You can specify the node name for a pod

Reply:
This is a typo, fixed to `<pod-node>`.

I am adapting to Gherkin Syntax, `< >` is the way presenting the scenario is running multiple times at different values. To specify the node name would duplicate the steps of the skeleton.

The values are matching
```
    Examples:        | pod-node | duplicate-node |        | node-1   | node-2         |        | node-2   | node-1         |        | node-1   | node-2         |
```